go/libraries/doltcore/doltdb/root_val.go

```diff
@@ -1149,6 +1178,60 @@ func (root *RootValue) PutForeignKeyCollection(ctx context.Context, fkc *Foreign
 	return &RootValue{root.vrw, rootValSt, fkc.copy()}, nil
 }
 
+// VerifyForeignKeys ensures that all foreign keys' tables are present, removing any foreign keys where the declared
+// table is missing, and returning an error if a key is in an invalid state or a referenced table is missing.
+func (root *RootValue) VerifyForeignKeys(ctx context.Context) (*RootValue, error) {
+	fkCollection, err := root.GetForeignKeyCollection(ctx)
+	if err != nil {
+		return nil, err
+	}
+	allTablesSlice, err := root.GetTableNames(ctx)
+	if err != nil {
+		return nil, err
+	}
+	allTablesSet := make(map[string]schema.Schema)
+	for _, tableName := range allTablesSlice {
+		tbl, ok, err := root.GetTable(ctx, tableName)
+		if err != nil {
+			return nil, err
+		}
+		if !ok {
+			return nil, fmt.Errorf("found table `%s` in staging but could not load for foreign key check", tableName)
+		}
+		tblSch, err := tbl.GetSchema(ctx)
+		if err != nil {
+			return nil, err
+		}
+		allTablesSet[tableName] = tblSch
+	}
+
+	// some of these checks are sanity checks and should never happen
+	allForeignKeys := fkCollection.AllKeys()
+	for _, foreignKey := range allForeignKeys {
+		tblSch, existsInRoot := allTablesSet[foreignKey.TableName]
+		if existsInRoot {
+			if !foreignKey.VerifyTableSchema(tblSch) {
+				return nil, fmt.Errorf("foreign key `%s` has entered an invalid state, table `%s` does not match", foreignKey.Name, foreignKey.TableName)
+			}
+			parentSch, existsInRoot := allTablesSet[foreignKey.ReferencedTableName]
+			if !existsInRoot {
+				return nil, fmt.Errorf("foreign key `%s` requires the referenced table `%s`", foreignKey.Name, foreignKey.ReferencedTableName)
+			}
+			if !foreignKey.VerifyReferencedTableSchema(parentSch) {
+				return nil, fmt.Errorf("foreign key `%s` has entered an invalid state, referenced table `%s` does not match",
```

Comment:
This needs to be a more specific error message. Spell out what the foreign key constraint expects